### PR TITLE
Display build version of workers in WebUI and capacity command

### LINKED ISSUE #13763: Add more information when registering the worker
**Is your feature request related to a problem? Please describe.**
When upgrading Alluxio in our cluster, we typically do a rolling upgrade in our K8s cluster. During that, we typically upgrade the workers in batches, so some workers in the cluster will have different version/revision from others. We would like that to be reported to the master so the master can display those information on the UI and in `alluxio fsadmin` commands. 

It will also be very helpful if we can associate issues/performance with the worker version/revision, so during the rolling upgrade we can be notified if the upgrade is bad and we should roll back.

Especially when we are using `SNAPSHOT` versions, it is hard to tell which version/revision the worker is on.

**Describe the solution you'd like**
We want to get more information about workers on the master side. Add the version and revision information of the worker

**Urgency**
Normal

**Additional context**
No

EDIT: rephrased use case and requirements on 2021/10/27

### PR DESCRIPTION
**Pull request by kaijchen**
### What changes are proposed in this pull request?

1. Add version and revision information in workerRegister proto and API. 
2. Add worker version and revision in `alluxio fsadmin report capacity`.
3. Add worker version and revision in workers tab of WebUI.

Fix #13763.

### Why are the changes needed?

1. When performing rolling upgrades, we need to know the status of the cluster.
(e.g. Which worker is still running the old version? Is recent upgrade sucessful?)
2. It's cumbersome to query each worker individually.
(And the WebUI of workers may not be exposed to the client due to firewall policy)
3. The revision is for distinguishing different builds of the same SNAPSHOT version.
(Useful for development and testing)

### Does this PR introduce any user facing changes?

1. Add worker version and revision in `alluxio fsadmin report capacity`.
```console
$ ./bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 1804B
        Tier: MEM  Size: 1804B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM              Version          Revision                                
9.134.131.52     0                capacity      20.74GB          2.9.0-SNAPSHOT   ecafed558c467208da86d23829c589d22634561c
                                  used          1804B (0%)
```
2. Add worker version and revision in workers tab of WebUI.
image

### CODE REVIEW

**Comment by kaijchen** [comment]
UI change example:

image


```console
$ ./bin/alluxio fsadmin report workers
Alluxio workers: 
    WorkerInfo{id=1070416111312515592, address=WorkerNetAddress{host=9.134.131.52, containerHost=, rpcPort=29999, dataPort=29999, webPort=30000, domainSocketPath=, tieredIdentity=TieredIdentity(node=9.134.131.52, rack=null)}, lastContactSec=0, state=In Service, capacityBytes=22264296789, usedBytes=1804, startTimeMs=1668048392202, capacityBytesOnTiers={MEM=22264296789}, usedBytesOnTiers={MEM=1804}, version=2.9.0-SNAPSHOT, revision=d44b43ffd5d38b414cfee5360610ad2f4b1f0829}
```

**Comment by kaijchen** [comment]
Cc @jiacheliu3 @apc999 PTAL.

**Comment by maobaolong** [comment]
@kaijchen Would you please add more description to show why we need to display the revision and worker version

**Comment by kaijchen** [comment]
> @kaijchen Would you please add more description to show why we need to display the revision and worker version

Thanks for the suggestion, I have updated the description. Feel free to ask more questions if it's not clear enough.

**Comment by kaijchen** [comment]
I have updated this PR based on the discussion with @jiacheliu3 in community meeting. The changes are:

1. Include BuildVersion in Options, and avoid the API change.
2. Include version info in `alluxio fsadmin report capacity`, and avoid adding a new command.

PR discription has been updated.

**Review by maobaolong — APPROVED** [review]
This feature looks great, we can get the version of workers through this command tool and web ui. Thank you for your contribution!

**Review by tieujason330 — APPROVED** [review]
ui changes LGTM, thanks

File: core/server/master/src/main/java/alluxio/master/block/meta/MasterWorkerInfo.java [hunk 1]
```diff
@@ -164,6 +168,7 @@ public MasterWorkerInfo(long id, WorkerNetAddress address) {
     mBlocks = new LongOpenHashSet();
     mToRemoveBlocks = new LongOpenHashSet();
     mLastUpdatedTimeMs = new AtomicLong(CommonUtils.getCurrentMs());
+    mBuildVersion = new AtomicReference<>(BuildVersion.getDefaultInstance());
```
Comment (dbw9580) [inline core/server/master/src/main/java/alluxio/master/block/meta/MasterWorkerInfo.java:171]: add this to the `toString` output.

**Review by dbw9580 — CHANGES_REQUESTED** [review]
Mostly LGTM with a few minor comments, thanks!

**Comment by kaijchen** [comment]
Thanks @dbw9580 for the suggestion, I have addressed those issues.
PS: The javadoc of revision is originally copied from
https://github.com/Alluxio/alluxio/blob/a95ed9a15bb5c96c85342748214115b6e41e4ad0/core/common/src/main/java-templates/alluxio/ProjectConstants.java#L20-L21

**Review by dbw9580 — APPROVED** [review]

**Review by jiacheliu3 — COMMENTED** [review]
One last comment from me is on back/forward compatibility. By using `optional` gRPC field and `optins.hasBuildVersion()` I think old/new master can work with old/new workers. But does old/new CapacityCommand work with old/new worker?

All other changes LGTM and we can merge once we confirm on the above question. Thanks a lot for the good work!

**Comment by kaijchen** [comment]
Thanks @jiacheliu3 for your questions.
I have tested v2.9.0 as old master/client with the current branch (e9e154f):

## New Master

### New Client

```console
$ bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 1804B
        Tier: MEM  Size: 1804B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM              Version          Revision                                
9.134.131.52     0                capacity      20.74GB          2.9.0-SNAPSHOT   8ae79c38ef6340a98665f39a96a305cbc1042bc7
                                  used          1804B (0%)
```

### Old Client

```console
$ bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 1804B
        Tier: MEM  Size: 1804B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM
9.134.131.52     0                capacity      20.74GB
                                  used          1804B (0%)
```

## Old Master

### New Client

```console
$ bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 0B
        Tier: MEM  Size: 0B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM              Version          Revision                                
9.134.131.52     0                capacity      20.74GB                                                                   
                                  used          0B (0%) 
```

### Old Client

```console
$ bin/alluxio fsadmin report capacity
bin/alluxio fsadmin report capacity
Capacity information for all workers: 
    Total Capacity: 20.74GB
        Tier: MEM  Size: 20.74GB
    Used Capacity: 0B
        Tier: MEM  Size: 0B
    Used Percentage: 0%
    Free Percentage: 100%

Worker Name      Last Heartbeat   Storage       MEM
9.134.131.52     0                capacity      20.74GB
                                  used          0B (0%)
```

**Comment by kaijchen** [comment]
We **can** check `hasBuildVersion` in `GrpcUtils#fromProto()`, however, there isn't any `hasField` check for other optional fields either.

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks!

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by kaijchen** [comment]
> We **can** check `hasBuildVersion` in `GrpcUtils#fromProto()`, however, there isn't any `hasField` check for other optional fields either.

@jiacheliu3 The new default value is correct is because of we are using protobuf3 to compile our proto2 file. If using protobuf 2 to compile it, there will be problem.

**Pull request merged**